src/main/java/net/earthcomputer/clientcommands/task/RenderDistanceScanTask.java

```diff
@@ -16,29 +18,62 @@
 import net.minecraft.world.level.chunk.LevelChunkSection;
 import net.minecraft.world.level.chunk.status.ChunkStatus;
 import net.minecraft.world.phys.Vec3;
+import org.jetbrains.annotations.Nullable;
 
-import java.util.Iterator;
+import java.lang.ref.WeakReference;
+import java.util.Set;
 import java.util.function.Predicate;
 
 public abstract class RenderDistanceScanTask extends SimpleTask {
     private static final long MAX_SCAN_TIME = 30_000_000L; // 30ms
+    private static final Set<Object> MUTEX_KEYS = Set.of(RenderDistanceScanTask.class);
+
+    static {
+        ClientLevelEvents.CHUNK_UPDATE.register((level, pos, oldState, newState) -> {
+            WeakReference<ClientLevelEvents.ChunkUpdate> chunkUpdateCallback = RenderDistanceScanTask.chunkUpdateCallback;
+            if (chunkUpdateCallback != null) {
+                ClientLevelEvents.ChunkUpdate callback = chunkUpdateCallback.get();
+                if (callback != null) {
+                    callback.onBlockStateUpdate(level, pos, oldState, newState);
+                }
+            }
+        });
+    }
 
-    private final boolean keepSearching;
-
-    private Iterator<BlockPos.MutableBlockPos> squarePosIterator;
+    @Nullable
+    private static WeakReference<ClientLevelEvents.ChunkUpdate> chunkUpdateCallback = null;
+    protected boolean keepSearching;
+    private LongLinkedOpenHashSet remainingChunks;
 
     protected RenderDistanceScanTask(boolean keepSearching) {
         this.keepSearching = keepSearching;
     }
 
     @Override
     public void initialize() {
-        squarePosIterator = createIterator();
+        remainingChunks = new LongLinkedOpenHashSet();
+        Entity cameraEntity = Minecraft.getInstance().cameraEntity;
+        if (cameraEntity == null) {
+            _break();
+            return;
+        }
+        BlockPos.spiralAround(new BlockPos(Mth.floor(cameraEntity.getX()) >> 4, 0, Mth.floor(cameraEntity.getZ()) >> 4), Minecraft.getInstance().options.renderDistance().get(), Direction.EAST, Direction.SOUTH).iterator().forEachRemaining(pos -> remainingChunks.add(ChunkPos.asLong(pos.getX(), pos.getZ())));
+        ClientLevelEvents.UNLOAD_CHUNK.register((level, pos) -> remainingChunks.remove(pos.toLong()));
+        ClientLevelEvents.LOAD_CHUNK.register((level, pos) -> remainingChunks.add(pos.toLong()));
```

Comment:
These events should be registered in the static initializer too. The weak reference should just be a reference to the task, keep it simple.